### PR TITLE
Fold "FrozenObjectHandle(REF) + CNS" to a byref constant

### PR DESCRIPTION
**Pull request by EgorBo**
Let's see if any target complains about it. This removes two "add" instructions from https://github.com/dotnet/runtime/pull/85880 snippet, or e.g.:
```csharp
ReadOnlySpan<char> Test() => "Hello World"
```
was:
```asm
; Method Program:Test():System.ReadOnlySpan`1[ushort]:this
       mov      rax, 0x14380209280      ; 'Hello World'
       add      rax, 12 ; offset to string's data
       mov      edx, 11 ; Length
       ret      
; Total bytes of code: 20
```
now:
```asm
; Method Program:Test():System.ReadOnlySpan`1[ushort]:this
       mov      rax, 0x29C8020928C ;;  "interior pointer"
       mov      edx, 11 ; Length
       ret      
; Total bytes of code: 16
```

[Diffs.](https://dev.azure.com/dnceng-public/public/_build/results?buildId=265318&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

File: src/coreclr/jit/lower.cpp [hunk 1]
```diff
@@ -6247,6 +6247,22 @@ GenTree* Lowering::LowerAdd(GenTreeOp* node)
...
+        // Fold "FrozenObjectHandle(REF) + CNS" to a byref constant:
+        //
+        //  *  ADD       byref
+        //  +--*  CNS_INT(h) ref
+        //  \--*  CNS_INT    long
+        //
+        // We can do this earlier but that will need some efforts (e.g. to restore original object from a
+        // byref constant for optimizations, be careful with "base constant" CSEs, etc). Also, it can't be
+        // just enabled for AOT, it needs some way to use reloc + offset then.
+        if (!comp->opts.IsReadyToRun() && op1->IsIconHandle(GTF_ICON_OBJ_HDL) && op2->IsCnsIntOrI())
```
Comment (jakobbotsch) [inline src/coreclr/jit/lower.cpp:6259]:
```suggestion
        if (!comp->opts.compReloc && op1->IsIconHandle(GTF_ICON_OBJ_HDL) && op2->IsCnsIntOrI())
```
?

Comment (EgorBo) [reply src/coreclr/jit/lower.cpp]: @jakobbotsch ah I've already applied a different suggesion 🙂

**Comment by EgorBo** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=265208&view=ms.vss-build-web.run-extensions-tab)

**Comment by EgorBo** [comment]
runtime-coreclr outerloop, runtime-coreclr jitstress, runtime-coreclr jitstressregs, runtime-coreclr gcstress0x3-gcstress0xc

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL simple change with [nice diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=266052&view=ms.vss-build-web.run-extensions-tab), outerloop/jitstress/gcstress tests passed

**Comment by EgorBo** [comment]
@jakobbotsch PTAL

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**